### PR TITLE
fix: store.put is async and might fail

### PR DESCRIPTION
**Pull request by ftreesmilo**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[X] Bug fix
[ ] New feature
[ ] Other, please explain:

**What changes did you make? (Give an overview)**
store.put was not checking if there was an error writing to the store.
We would incorrectly mark that we had a piece, because it was never successfully written if an error was thrown.

**Which issue (if any) does this pull request address?**
n/a

**Is there anything you'd like reviewers to focus on?**

### CODE REVIEW

File: lib/torrent.js [hunk 1]
```diff
@@ -1599,29 +1599,33 @@ class Torrent extends EventEmitter {
...
+              self.pieces[index] = new Piece(piece.length)
+              self.bitfield.set(index, false)
+              self.select(index, index, 1)
+              self.emit('warning', err)
+            } else {
+              self.pieces[index] = null
+              self._markVerified(index)
+              self.wires.forEach(wire => {
+                wire.have(index)
+              })
```
Comment (ftreesmilo) [inline lib/torrent.js:1617]: @jhiesey The error path is one part, but the other big part of this change is marking the piece as verified when storage is done with the put.

**Comment by jhiesey** [comment]
@ftreesmilo sorry I've been busy. I promise I'll take a look at this tomorrow.

**Review by jhiesey — COMMENTED** [review]
We definitely want the backpressure fix! I'm less convinced about infinite retries, but I can write a retry wrapper if that will work for your use case @ftreesmilo

**Review by feross — APPROVED** [review]
Let's merge just the backpressure part for now, since that's definitely an issue and those changes definitely fix the problem. We just confirmed that it was affecting Wormhole and the fixes solve the problem.

That's the thing about backpressure – if you forgot to do it in even one place, then the whole pipeline can back up at that place.

Thanks a lot for catching this, @ftreesmilo!

**Comment by jhiesey** [comment]
@ftreesmilo let me know if you want me to make a `RetryChunkStore` or something like that.

File: lib/torrent.js [hunk 2]
```diff
@@ -1599,29 +1599,33 @@ class Torrent extends EventEmitter {
...
-
           self.store.put(index, buf, err => {
-            if (err) self._destroy(err)
-          })
-
-          self.wires.forEach(wire => {
-            wire.have(index)
+            if (err) {
+              // Not sure what else to do here if store threw up.
+              // try to download again?
```
[the remark this answers is not in the document]
Comment (ftreesmilo) [reply lib/torrent.js]: I think that's probably a good idea.

File: lib/torrent.js [hunk 3]
```diff
@@ -1599,29 +1599,33 @@ class Torrent extends EventEmitter {
...
+              self.pieces[index] = new Piece(piece.length)
+              self.bitfield.set(index, false)
+              self.select(index, index, 1)
```
[the remark this answers is not in the document]
Comment (ftreesmilo) [reply lib/torrent.js]: I wasn't sure about select... wanted to be sure.

**Comment by ftreesmilo** [comment]
@jhiesey If you want to take a stab at it, go for it :)
now that this is merged, I have to start unwinding my private fork.   I'm going to be busy for a bit.

**Pull request merged**